cachi2/core/package_managers/pip.py

```diff
@@ -1463,6 +1461,22 @@ def _split_hashes_from_options(options: list[str]) -> tuple[list[str], list[str]
         return hashes, reduced_options
 
 
+def _standardize_name(name: str) -> str:
+    return re.sub("[^A-Za-z0-9.]+", "-", name)
+
+
+def _standardize_version(version: str) -> str:
+    try:
+        return str(packaging.version.Version(version))
+    except packaging.version.InvalidVersion:
+        version = version.replace(" ", ".")
+        return re.sub("[^A-Za-z0-9.]+", "-", version)
```

Comment:
Same thing about vendoring with ^these 2 - part of the issue is investigating the proper replacements for these and by quickly going through packaging's docs [canonicalize_name](https://packaging.pypa.io/en/stable/utils.html#packaging.utils.canonicalize_name) and [version](https://packaging.pypa.io/en/stable/utils.html#packaging.utils.canonicalize_version) seem like candidates worth exploring whether they accomplish what we need.

Reply:
@ianrochapg I see you pushed a new revision, have you looked at the canonicalize functions I pointed to in my earlier comment to see if it worked and integration tests passed? If so, that would be obviously a win for us because we wouldn't need the 2 helpers you're introducing.

Reply:
After checking it out, we can use `version`, but `name` doesn't cover every scenario required.

Reply:
I wonder if we can use `canonicalize_name` alongside `canonicalize_version` from packaging...

Reply:
@slimreaper35 It does work, I just thought I didn't need to import something else since I do need `_standardize_name` for line 316 and it covers even more scenarios than `canonicalize_name`.

Reply:
I noticed there's actually a difference in behavior between `canonicalize_name` and `_standardize_name`.

```
>>> canonicalize_name("aio.http")
'aio-http'
>>> _standardize_name("aio.http")
'aio.http'
```

So, _standardize does allow for `.`, whereas canonicalize does not. 

In the end, for [requirement.package](https://github.com/containerbuildsystem/cachi2/blob/main/cachi2/core/package_managers/pip.py#L1260), I don't know if it causes any issue, since `canonicalize_name` will be called regardless for PyPI deps (see [1](https://github.com/containerbuildsystem/cachi2/blob/main/cachi2/core/package_managers/pip.py#L1879), [2](https://github.com/jwodder/pypi-simple/blob/master/src/pypi_simple/client.py#L278), [3](https://github.com/jwodder/pypi-simple/blob/master/src/pypi_simple/client.py#L10)). 

I did not dig deeper to see if it impacts file or git dependencies, but if it does not, we could just keep using `req.name`.  Otherwise, to keep the previous behavior, we want `_standardize_name`.

Reply:
The only problem that we have currently is this line (otherwise, we wouldn't need `_standardize` at all):
```
@pytest.mark.parametrize("repo_name_with_subpath", ["bar-subpath", None])
```
`/` are replaced with `-`

Is it the same for all package managers?

Reply:
It is definitely not the same for gomod, as the module's name always contains a `/`. Not sure if npm/yarn allow for `/` in the package's name. Maybe we should just keep the normalize for this specific [instance](https://github.com/ianrochapg/cachi2/blob/pip-requirement-subclass/cachi2/core/package_managers/pip.py#L316), and keep the current behavior.

I'm still not sure if we can completely remove the _standardize from the other [instance](https://github.com/ianrochapg/cachi2/blob/pip-requirement-subclass/cachi2/core/package_managers/pip.py#L1314). Keeping it will maintain the same behavior prior to this PR.

Reply:
> I'm still not sure if we can completely remove the _standardize from the other [instance](https://github.com/ianrochapg/cachi2/blob/pip-requirement-subclass/cachi2/core/package_managers/pip.py#L1314). Keeping it will maintain the same behavior as the current code has.

As I mentioned above, we can replace it with `canonicalize_name` from the packaging package.

Well, I don't know then, I can see `/` in the rpm SBOM too.

Reply:
> As I mentioned above, we can replace it with canonicalize_name from the packaging package.
Yes, that would work, but we wouldn't be able to completely get rid of `_standardize_name` since it is used to normalize package names that are resolved from a URL.

Reply:
I admit I got lost in this thread quite a bit already (I have switched the context a long time ago). So I'll need some refreshment and more explanation I guess in layman terms on why do we want to stick to `_standardize_name` when that regex isn't even [PEP 503](https://peps.python.org/pep-0503/#normalized-names) compatible? Conversely, `canonicalize_name` is [1].

So what am I missing here in terms of normalization?  While the proposed behaviour is in check with the old `pkg_resources` 's `safe_name` [2] behaviour it doesn't adhere to PEP. Shouldn't we strive all Python code we write to conform to PEP standards?

[1]  https://github.com/pypa/packaging/blob/bb61b7bcbc429b49e6659a8b6c2966b6ebd046b6/src/packaging/utils.py#L48C26-L48C33
[2] https://github.com/pypa/setuptools/blob/main/pkg_resources/__init__.py#L1550

Reply:
There are two problems that got tangled here, and both relate to the usage of `_standardize_name`:

1- we [use](https://github.com/ianrochapg/cachi2/blob/pip-requirement-subclass/cachi2/core/package_managers/pip.py#L316) it to normalize the name of the main package (the one that is being processed). We need to keep it here since `canonicalize_name` allows `/` in the name, and the current behavior is to replace `/` to `-`. `/` is a valid use case here because the name can be inferred from the repository's URL.

2- we also [use](https://github.com/ianrochapg/cachi2/blob/pip-requirement-subclass/cachi2/core/package_managers/pip.py#L1314) it to normalize the other package's names (dependencies). We can definitely use `canonicalize_name` here, since it covers all use cases.

Ian opted to keep both instances of `_standardize_name` since we wouldn't be able to get rid of it, but I agree we can change the second instance to `canonicalize_name`.

We could probably change the current behavior and allow for `/` in the main package name, but I personally wouldn't want to go this far, at least not in this PR.

Reply:
Also, PEP 503 targets Simple Repository API, and in case "1" above, we're talking about how the name will be presented in the resulting SBOM.

Reply:
> 1- we [use](https://github.com/ianrochapg/cachi2/blob/pip-requirement-subclass/cachi2/core/package_managers/pip.py#L316) it to normalize the name of the main package (the one that is being processed). We need to keep it here since canonicalize_name allows / in the name, and the current behavior is to replace / to -. / is a valid use case here because the name can be inferred from the repository's URL.


> Also, PEP 503 targets Simple Repository API, and in case "1" above, we're talking about how the main component name will be presented in the resulting SBOM.

Now I see @brunoapimentel . So, how about instead of inventing our own function for a single use case, we additionally filter out `/` from the name we get out of canonicalize to retain the current behavior wrt/ SBOM?

Reply:
I'm ok with that solution :+1: 

I'm sure that doing `canonicalize + replace '/'` will make all tests pass, but I think it's best to double check if we're not uncovering any other corner case with this change. 

(sorry, the detailed info I had about this was already garbage collected :brain: :rofl:)

Reply:
So, in reality, there's a bunch of characters that would get the '-' treatment from the old normalization and not canonicalize, like e.g. '@' , then the question is whether that's a valid character and PEP says it is, so I guess like with anything we can go with canonicalize provided existing tests pass and fix other occurrences on the fly if needed, it's an easy fix.

Reply:
@brunoapimentel @eskultety Per your suggestion, I replaced `standardize_name` with `canonicalize_name` while also replacing the `/` with a `-` .